### PR TITLE
[Tables] Client and pipeline refactor

### PR DESCRIPTION
**Pull request by annatisch**

### CODE REVIEW

File: sdk/tables/azure-data-tables/azure/data/tables/_base_client.py [hunk 1]
```diff
@@ -33,74 +30,63 @@
...
-        service,  # type: str
+        account_url,  # type: Any
         credential=None,  # type: Optional[Any]
         **kwargs  # type: Any
     ):
         # type: (...) -> None
-        self._location_mode = kwargs.get("_location_mode", LocationMode.PRIMARY)
+        try:
+            if not account_url.lower().startswith("http"):
+                account_url = "https://" + account_url
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/azure/data/tables/_base_client.py:68]:
```suggestion
            if not account_url.lower().startswith("http") and not account_url.lower.startswith("https"):
```

File: sdk/tables/azure-data-tables/azure/data/tables/_base_client.py [hunk 2]
```diff
@@ -125,48 +111,12 @@ def __init__(
...
-
-    def __exit__(self, *args):
-        self._client.__exit__(*args)
-
-    def close(self):
-        """This method is to close the sockets opened by the client.
-        It need not be used when using with a context manager.
-        """
-        self._client.close()
-
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/azure/data/tables/_base_client.py:169]: Why did you get rid of these?

File: sdk/tables/azure-data-tables/azure/data/tables/_table_service_client.py [hunk 3]
```diff
@@ -370,9 +356,6 @@ def get_table_client(self, table_name, **kwargs):
             self.url,
             table_name=table_name,
             credential=self.credential,
-            key_resolver_function=self.key_resolver_function,
-            require_encryption=self.require_encryption,
-            key_encryption_key=self.key_encryption_key,
             api_version=self.api_version,
             transport=self._client._client._pipeline._transport,  # pylint: disable=protected-access
             policies=self._policies,
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/azure/data/tables/_table_service_client.py:361]: Discussed this out of band a few days ago, are we using the `_pipeline` we build a few lines before this.

File: sdk/tables/azure-data-tables/azure/data/tables/aio/_base_client_async.py [hunk 4]
```diff
@@ -95,38 +80,22 @@ def _configure_credential(self, credential):
...
+            self._credential_policy,
+            CustomHookPolicy(**kwargs),
+            NetworkTraceLoggingPolicy(**kwargs),
             DistributedTracingPolicy(**kwargs),
             HttpLoggingPolicy(**kwargs),
         ]
 
-        if self._cosmos_endpoint:
-            self._policies.insert(0, CosmosPatchTransformPolicy())
-
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/azure/data/tables/aio/_base_client_async.py:129]: Why was this removed?

File: sdk/tables/azure-data-tables/azure/data/tables/aio/_base_client_async.py [hunk 5]
```diff
@@ -28,44 +17,41 @@
...
+        # type: (...) -> None
+        super(AsyncTablesBaseClient, self).__init__(account_url, credential=credential, **kwargs)
+        self._client = AzureTable(
+            self.url,
+            policies=kwargs.pop('policies', self._policies),
+            **kwargs
+        )
 
     async def __aenter__(self):
         await self._client.__aenter__()
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/azure/data/tables/aio/_base_client_async.py:57]: Because this inherits from the sync base client, will there be a weird behavior where you can use an async client as a sync context manager:
```python
with TableClient(...) as tc:
    await tc.get_entity(...)
```

File: sdk/tables/azure-data-tables/azure/data/tables/aio/_policies_async.py [hunk 6]
```diff
@@ -3,317 +3,153 @@
...
-            if asyncio.iscoroutine(response_callback):
-                await response_callback(response)
-            else:
-                response_callback(response)
-            request.context["response_callback"] = response_callback
-        return response
+    :keyword RetryMode retry_mode: Fixed or exponential delay between attemps, default is exponential.
 
+    :keyword int timeout: Timeout setting for the operation in seconds, default is 604800s (7 days).
+    """
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/azure/data/tables/aio/_policies_async.py:52]: Just for context, why is the timeout setting so high?

File: sdk/tables/azure-data-tables/azure/data/tables/aio/_table_service_client_async.py [hunk 7]
```diff
@@ -400,9 +379,6 @@ def get_table_client(
             self.url,
             table_name=table_name,
             credential=self.credential,
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/azure/data/tables/aio/_table_service_client_async.py:381]: same question about the `_pipeline`

File: sdk/tables/azure-data-tables/azure/data/tables/aio/_base_client_async.py (hunk 4 above)
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/azure/data/tables/aio/_base_client_async.py:129]: Nevermind, I found this in a different place.

File: sdk/tables/azure-data-tables/azure/data/tables/_table_service_client.py (hunk 3 above)
Comment (annatisch) [reply sdk/tables/azure-data-tables/azure/data/tables/_table_service_client.py]: I hadn't _planned_ to fix that in this PR - however I ended up needing to for the retry tests - so this should be be resolved :)

File: sdk/tables/azure-data-tables/azure/data/tables/_policies.py [hunk 8]
```diff
@@ -4,225 +4,54 @@
...
-import re
-import random
-from time import time
-from io import SEEK_SET, UnsupportedOperation
-import logging
-import uuid
-import types
-from typing import Any, TYPE_CHECKING
+import time
+from typing import Any, TYPE_CHECKING, Dict
```
[the remark this answers is not in the document]
Comment (annatisch) [reply sdk/tables/azure-data-tables/azure/data/tables/_policies.py:8]: Nah - this should be fixed in pylint now.

File: sdk/tables/azure-data-tables/azure/data/tables/aio/_base_client_async.py (hunk 5 above)
Comment (annatisch) [reply sdk/tables/azure-data-tables/azure/data/tables/aio/_base_client_async.py:59]: Nope - the async client does _not_ inherit from the sync base client. For exactly this reason. It only inherits from the AccountHostsMixin

File: sdk/tables/azure-data-tables/azure/data/tables/aio/_policies_async.py (hunk 6 above)
Comment (annatisch) [reply sdk/tables/azure-data-tables/azure/data/tables/aio/_policies_async.py:52]: hmm good question - but it's a question for azure-core - as I just copying the docstring from the base retry policy (which is what this retry policy inherits from).

File: sdk/tables/azure-data-tables/azure/data/tables/aio/_table_service_client_async.py (hunk 7 above)
Comment (annatisch) [reply sdk/tables/azure-data-tables/azure/data/tables/aio/_table_service_client_async.py:380]: Fixed :)

File: sdk/tables/azure-data-tables/tests/test_retry.py [hunk 9]
```diff
@@ -97,91 +96,66 @@ def test_retry_on_server_error(self, tables_storage_account_name, tables_primary
...
-    #     try:
-    #         with pytest.raises(HttpResponseError) as error:
-    #             self.ts.create_table(new_table_name, raw_response_hook=callback)
-    #         self.assertEqual(error.value.response.status_code, 408)
-    #         self.assertEqual(error.value.reason, 'Created')
+        try:
+            with pytest.raises(HttpResponseError) as error:
+                self.ts.create_table(new_table_name, raw_response_hook=callback)
+            self.assertEqual(error.value.response.status_code, 500)
+            self.assertEqual(error.value.reason, 'Created')
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/tests/test_retry.py:155]:
```suggestion
             assert error.value.response.status_code == 500
             assert error.value.reason == 'Created'
```
Nit: trying to keep `unittest` out of tests so we can eventually remove it.

File: sdk/tables/azure-data-tables/tests/test_retry_async.py [hunk 10]
```diff
@@ -95,93 +95,66 @@ async def test_retry_on_server_error_async(self, tables_storage_account_name, ta
...
-    #     try:
-    #         with with pytest.raises(HttpResponseError) as error:
-    #             await self.ts.create_table(new_table_name, raw_response_hook=callback)
-    #         self.assertEqual(error.value.response.status_code, 408)
-    #         self.assertEqual(error.value.reason, 'Created')
+        try:
+            with pytest.raises(HttpResponseError) as error:
+                await self.ts.create_table(new_table_name, raw_response_hook=callback)
+            self.assertEqual(error.value.response.status_code, 500)
+            self.assertEqual(error.value.reason, 'Created')
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/tests/test_retry_async.py:154]:
```suggestion
            assert error.value.response.status_code == 500
            assert error.value.reason == 'Created'
```
same nit

**Review by seankane-msft — COMMENTED** [review]
A few small things

File: sdk/tables/azure-data-tables/azure/data/tables/_base_client.py [hunk 11]
```diff
@@ -33,74 +28,78 @@
...
+    # type: (Dict[str, Any], str) -> str
+    api_version = kwargs.pop("api_version", None)
+    if api_version and api_version not in _SUPPORTED_API_VERSIONS:
+        versions = "\n".join(_SUPPORTED_API_VERSIONS)
+        raise ValueError(
+            "Unsupported API version '{}'. Please select from:\n{}".format(
+                api_version, versions
+            )
+        )
+    return api_version or default
```
[the remark this answers is not in the document]
Comment (annatisch) [reply sdk/tables/azure-data-tables/azure/data/tables/_base_client.py:68]: One of the guidelines is that all SDKs support an API version parameter. So that part is mandatory. However in next stand up we should confirm what all languages are setting as the default API version.

**Review by seankane-msft — APPROVED** [review]
LGTM :shipit:
